taca/delivery/delivery_classes.py

```diff
@@ -111,3 +141,318 @@ def stage_data(self):
+                f"to DDS delivery project: {', '.join(other_member_emails)}"
+            )
+            for email in other_member_emails:
+                if email not in other_member_details:
+                    other_member_details.append(email)
+        return other_member_details
+
+    def get_project_description(self, given_desc=None):
+        """Set project description, either given or from order portal"""
+        if given_desc:
+            logger.warning(
+                f"Project description for project {self.project_id} specified by user: {given_desc}"
+            )
+            return given_desc
+        else:
+            project_name = self.order_details.get("fields", {}).get("project_ngi_name")
+            created_desc = f"{project_name} ({datetime.now().strftime('%Y-%m-%d')})"
+            logger.info(
+                f"Project description for project {self.project_id}: {created_desc}"
+            )
+            return created_desc
+
+    def create_dds_project(self):
+        """Create a DDS delivery project and return the ID."""
+        dds_command = [
+            "dds",
+            "--no-prompt",
+            "project",
+            "create",
+            "--title",
+            self.project_id,
+            "--description",
+            self.project_description,
+            "--principal-investigator",
+            self.pi_email,
+            "--owner",
+            self.pi_email,
+        ]
+        if self.other_member_details:
+            for member in self.other_member_details:
+                dds_command.append(f"--researcher {member}")
+        dds_project_id = ""
+        try:
+            output = ""
+            for line in self._execute(dds_command):
+                output += line
+                print(line, end="")
+        except subprocess.CalledProcessError as e:
+            logger.exception(
+                "An error occurred while setting up the DDS delivery project."
+            )
+            raise e
+        project_pattern = re.compile("ngisthlm\d{5}")
+        found_project = re.search(project_pattern, output)
+        if found_project:
+            dds_project_id = found_project.group()
+            return dds_project_id
+        else:
+            raise AssertionError(f"DDS project NOT set up for {self.project_id}")
+
+    def upload_data(self, delivery_id):
+        """Upload staged data with DDS"""
+        log_dir = os.path.join(
+            os.path.dirname(CONFIG.get("log").get("file")), "DDS_logs"
+        )
+        project_log_dir = os.path.join(log_dir, self.project_id)
+        cmd = [
+            "dds",
+            "--no-prompt",
+            "data",
+            "put",
+            "--mount-dir",
+            project_log_dir,
+            "--project",
+            delivery_id,
+            "--source",
+            self.stage_dir,
+        ]
+        try:
+            output = ""
+            for line in self._execute(cmd):
+                output += line
```

Comment:
Does this generate a lot of output? I'm a bit worried it could accumulate or become a performance bottleneck

Reply:
It's so that we can see the progress bars that DDS prints. It's the same as we use today in taca-ngi-pipeline and it's been working OK so far I think.